File: Python/displayed_map.py
import pygame
from player import Player

def divisors(n: int):
    return [i for i in range(1, n + 1) if n % i == 0]

def matrix_extractor(matrix: list[list], x: int, y: int, size: int):
    return [[matrix[y + j][x + i] for i in range(size)] for j in range(size)]

class DisplayedMap:
    def __init__(self, map: list[list], map_center: int, screen_size: int):
        self.map = map

        self.divisors_tab = divisors(screen_size)
        self.zoom_lvl = 3

        self.n_cells = self.divisors_tab[self.zoom_lvl]
        self.cell_size = self.divisors_tab[-self.zoom_lvl-1]

        self.x = self.y = map_center - self.n_cells // 2

        self.grid = matrix_extractor(self.map, self.x, self.y, self.n_cells)
    
    def update_grid(self):
        self.grid = matrix_extractor(self.map, self.x, self.y, self.n_cells)
    
    def zoom(self, in_or_out: str, player: Player):
        if in_or_out == "out" and self.zoom_lvl < 7:
            self.zoom_lvl += 1
        elif in_or_out == "in" and self.zoom_lvl > 3:
            self.zoom_lvl -= 1
        
        self.n_cells = self.divisors_tab[self.zoom_lvl]
        self.cell_size = self.divisors_tab[-self.zoom_lvl - 1]

        half_size = self.n_cells // 2 # half the size of the displayed map

        self.x = max(1, min(128 - self.n_cells, player.absolute_x - half_size))
        self.y = max(1, min(128 - self.n_cells, player.absolute_y - half_size))

        player_relative_x = max(-half_size, min(half_size, player.absolute_x - (self.x + half_size)))
        player_relative_y = max(-half_size, min(half_size, player.absolute_y - (self.y + half_size)))

        return player_relative_x, player_relative_y
    
    def draw(self, screen):
        for i in range(len(self.grid)):
            row = self.grid[i]
            for j in range(len(row)):
                cell = row[j]
                color = "green" if cell else "blue"
                cell_square = pygame.Rect(j * self.cell_size, i * self.cell_size, self.cell_size, self.cell_size)
                pygame.draw.rect(screen, color, cell_square) # Grid squares
                pygame.draw.rect(screen, "black", cell_square, 1) # Grid lines

File: displayed_map.cpp
#include "displayed_map.h"
#include "player.h"
#include <algorithm>
#include <cmath>

std::vector<int> DisplayedMap::divisors(int n) {
    std::vector<int> result;
    for (int i = 1; i <= n; i++) {
        if (n % i == 0) {
            result.push_back(i);
        }
    }
    return result;
}

std::vector<std::vector<int>> DisplayedMap::matrix_extractor(const std::vector<std::vector<int>>& matrix, int x, int y, int size) {
    std::vector<std::vector<int>> result;
    for (int j = 0; j < size; j++) {
        std::vector<int> row;
        for (int i = 0; i < size; i++) {
            row.push_back(matrix[y + j][x + i]);
        }
        result.push_back(row);
    }
    return result;
}

DisplayedMap::DisplayedMap(const std::vector<std::vector<int>>& map, int map_center, int screen_size) 
    : map(map), zoom_lvl(3) {
    
    divisors_tab = divisors(screen_size);
    n_cells = divisors_tab[zoom_lvl];
    cell_size = divisors_tab[divisors_tab.size() - zoom_lvl - 1];
    
    x = y = map_center - n_cells / 2;
    
    grid = matrix_extractor(map, x, y, n_cells);
}

void DisplayedMap::update_grid() {
    grid = matrix_extractor(map, x, y, n_cells);
}

std::pair<int, int> DisplayedMap::zoom(const std::string& in_or_out, Player& player) {
    if (in_or_out == "out" && zoom_lvl < 7) {
        zoom_lvl++;
    } else if (in_or_out == "in" && zoom_lvl > 3) {
        zoom_lvl--;
    }
    
    n_cells = divisors_tab[zoom_lvl];
    cell_size = divisors_tab[divisors_tab.size() - zoom_lvl - 1];
    
    int half_size = n_cells / 2;
    
    x = std::max(1, std::min(128 - n_cells, player.get_absolute_x() - half_size));
    y = std::max(1, std::min(128 - n_cells, player.get_absolute_y() - half_size));
    
    int player_relative_x = std::max(-half_size, std::min(half_size, player.get_absolute_x() - (x + half_size)));
    int player_relative_y = std::max(-half_size, std::min(half_size, player.get_absolute_y() - (y + half_size)));
    
    return std::make_pair(player_relative_x, player_relative_y);
}

void DisplayedMap::draw(ALLEGRO_DISPLAY* display) {
    for (int i = 0; i < grid.size(); i++) {
        for (int j = 0; j < grid[i].size(); j++) {
            int cell = grid[i][j];
            ALLEGRO_COLOR color = cell ? al_map_rgb(0, 255, 0) : al_map_rgb(0, 0, 255);
            
            int x_pos = j * cell_size;
            int y_pos = i * cell_size;
            
            al_draw_filled_rectangle(x_pos, y_pos, x_pos + cell_size, y_pos + cell_size, color);
            al_draw_rectangle(x_pos, y_pos, x_pos + cell_size, y_pos + cell_size, al_map_rgb(0, 0, 0), 1);
        }
    }
} 